PowerEditor/src/NppBigSwitch.cpp

```diff
@@ -2059,14 +2059,28 @@ LRESULT Notepad_plus::process(HWND hwnd, UINT message, WPARAM wParam, LPARAM lPa
 		}
 
 		case WM_QUERYENDSESSION:
-		case WM_CLOSE:
 		{
-			if (message == WM_QUERYENDSESSION)
-			{
-				nppParam.queryEndSessionStart();
-			}
+			// app should return TRUE or FALSE immediately upon receiving this message,
+			// and defer any cleanup operations until it receives WM_ENDSESSION (with WPARAM TRUE)
+
+			// for a bigger tidy-up/save operations we can kick off a background thread here to prepare for shutdown
+			// and when we get the WM_END�SESSION TRUE, we wait there until that background operation completes
+			// before telling the system, "ok, you can shut down now...", i.e. returning 0 there
+
+			// whatever we do from here - make sure that it is ok for the operation to occur even if the shutdown
+			// is then subsequently canceled
 
-			if (nppParam.isQueryEndSessionStarted() && nppParam.doNppLogNulContentCorruptionIssue())
+			// here we could also display a prompt to ask the users whether they want to save their unsaved changes etc.,
+			// but in practice, this is usually not a good idea because if we do not respond to this message
+			// after a few seconds (e.g. user is away from PC...), the system will shut down without us
+
+			bool isFirstQueryEndSession = !nppParam.isEndSessionStarted();
+

```

Comment:
As MS win32 API manual for WM_QUERYENDSESSION's description, I would add a boolean here:
```cpp
bool isForcedShuttingDown = (lParam & ENDSESSION_CRITICAL);
```